pyxis_slurmstepd.c

```diff
@@ -251,6 +264,85 @@ static int enroot_create_user_runtime_dir(void)
 	return (0);
 }
 
+/* As root, create the per-user directory where squashfs files are stored and later could be reused. */
+static int enroot_create_image_save_dir(char *path) 
+{
+	int ret;
+	char uid_dir_path[PATH_MAX];
+
+	if (path == NULL || strlen(path) == 0) {
+		slurm_error("pyxis: path is empty");
+		return (-1);
+	}
+
+	/* In case it's a dir, we create per user folder */
+	if (path[strlen(path) - 1] == '/') {
+		ret = snprintf(uid_dir_path, sizeof(uid_dir_path), "%s%u/", path, context.job.uid);
+		if (ret < 0 || ret >= sizeof(uid_dir_path)) {
+			slurm_error("pyxis: too long path to save image: %s", path);
+			return (-1);
+		}
+		path = uid_dir_path;
+	}
+
+	/* Start from 1 to avoid first / */
+	for (size_t i = 1; i < strlen(path); i++) {
+		if (path[i] != '/') {
+			continue;
+		}
+
+		char temp_path[i + 1];
+		strncpy(temp_path, path, i);
+		temp_path[i] = '\0';
+
+		ret = mkdir(temp_path, 0700);
+		if (ret < 0 && errno != EEXIST) {
+			slurm_error("pyxis: couldn't mkdir %s: %s", temp_path, strerror(errno));
+			return (-1);
+		} else if (ret < 0 && errno == EEXIST) {
+			continue;
+		}
+
+		ret = chown(temp_path, context.job.uid, context.job.gid);
+		if (ret < 0) {
+			slurm_error("pyxis: couldn't chown %s: %s", temp_path, strerror(errno));
+			return (-1);
+		}
+	}
+
+	return (0);
+}
+
+static char* path_dir(const char *path)
```

Comment:
there is a `dirname` in libgen.h, but i'm not sure if it's easy to use it here

Reply:
you can also use `stdup` + `strrchr`
```
  char* dir_name = strdup(path);
  if (!dir_name) 
    return NULL;
  char* last = strrchr(dir_name, '/');
  if (last) {
      *(last+1) = '\0';
      return dir_name;
  }
  free(dir_name);
  return strdup("");
```

Reply:
looks like I don't need it anymore :)

Reply:
added back, I actually need this :D